### DISCUSSION TITLE
v2.0.0

### DISCUSSION THREAD

**Discussion by flyingpie** [flyingpie/windows-terminal-quake#119] [General]
# Introduction
The next major version of Windows-Terminal-Quake, with support for multiple apps simultaneously, including terminals other than Windows Terminal, and apps that are not terminals at all.

Example, with:

- Ctrl + 1: Windows Terminal
- Ctrl + 2: [Q-Dir](http://q-dir.com/)
- Ctrl + 3: [Process Hacker](https://processhacker.sourceforge.io/)
- Ctrl + 4: Spotify

- Ctrl + Q: Most recent app

https://github.com/flyingpie/windows-terminal-quake/assets/1295673/1bedd582-833a-4b7a-895a-f9d4de6d8ba7

# Current Releases
Portable downloads:
https://github.com/flyingpie/windows-terminal-quake/releases/tag/v2.0.0-pre3

Scoop:
TODO

WinGet:
```
winget install windows-terminal-quake-prerelease
```

# Plan
The current structure of WTQ (v1.x) has been built entirely on the presumption that there is only ever 1 instance of Windows Terminal. That makes it quite hard to add the ability to handle multiple applications, at the same time.

So I've started on a new application base that is more suited for this expansion, and will add the existing v1 functionality as I go. The new setup can be found under the [v2 branch](https://github.com/flyingpie/windows-terminal-quake/tree/v2).

# Builds
An early build can be downloaded from the [releases page](https://github.com/flyingpie/windows-terminal-quake/releases/tag/2.0.0-pre2).

Note that this build is missing most of the functionality of v1. The initial focus is around converting the app into one that can handle quake-style drop-down on multiple apps, that may or may not be terminal emulators.

# Configuration
Although v2 will be backwards compatible with v1 configuration files in the future, it is not yet.

Excerpt from the configuration file that is used with the video above (windows-terminal-quake.jsonc, included in the build):

```jsonc
{
  "Apps": [
    {
      "Name": "Terminal",
      "HotKeys": [ { "Modifiers": "Control", "Key": "1" } ],  // Hotkey specifically for this app.
      "FindExisting": {
        "ProcessName": "WindowsTerminal" // The name of the process to find, note that it's without ".exe".
      }
    },
    {
      "Name": "Spotify",
      "HotKeys": [ { "Modifiers": "Control", "Key": "4" } ],
      "FindExisting": {
        "ProcessName": "Spotify"
      }
    }
  ],
  "HotKeys": [ { "Modifiers": "Control", "Key": "Q" } ] // Hotkey that toggles the most recent app.
}
```

# Current Issues (non-exhaustive)
**Missing stuff**
V2 is currently missing lots of stuff from v1, while the new basic structure is being put in place to allow for multi-app support.

**No auto-start yet**
Quake'd applications **are currently not started by WTQ** v2, they need to be started manually. They can however be exited and re-started, without WTQ needing to restart.

**Big boi**
The build size (at ~160MB) is ridiculous. I'd like to explore self-contained aot+trimmed CoreCLR builds, but that does require getting rid of the Windows Forms dependency. Said dependency is used for the tray icon and the global hot key functionality.

~~**No tray icon yet**~~ (added in pre2)
The current build pops a command prompt, it doesn't drop to the tray yet. I'd like to investigate a different method than v1 used, since that relies on .Net WinForms.

# Future Stuff
- Automate different deployments types (GitHub release, Scoop, WinGet, etc.).
  - Some work done on builds, currently outputting zips for self-contained and framework-dependent versions

**Comment by fluxium** [comment]
Working well so far with https://github.com/alacritty/alacritty/releases/tag/v0.13.1

Super excited for the direction this is going! Thanks @flyingpie!

**Reply by flyingpie** [reply]
Thank you so much, I'm happy it's of use to you!

Feel free to drop a line if you encounter issues, or have suggestions :)

**Comment by flyingpie** [comment]
New release available, added tray icon, auto-hide, reduced flickering on some apps, and lots of invisible stuff.
https://github.com/flyingpie/windows-terminal-quake/releases/tag/2.0.0-pre2

**Reply by neondeex** [reply]
Nice! Does this reduce the q-dir flickering?

**Reply by flyingpie** [reply]
It does! :)

**Reply by flyingpie** [reply]
@neondeex Could you confirm or deny whether the flickering has reduced for you?

**Reply by neondeex** [reply]
@flyingpie Working super good with q-dir
but with notion im having problems that when i open it after some time ti is blank and i have to click to do the rendering

[![Image from Gyazo](https://i.gyazo.com/c8f0cb41b5881ec8792b1a6d7f9fa834.gif)](https://gyazo.com/c8f0cb41b5881ec8792b1a6d7f9fa834)

And partially fix the autostar just by having the apps autostart themself

**Comment by flyingpie** [comment]
We have a prerelease on WinGet now:

```
winget install windows-terminal-quake-prerelease
```

The biggest thing missing is an icon in the start menu or something, but small steps!

**Comment by flyingpie** [comment]
Prerelease 5 available:
https://github.com/flyingpie/windows-terminal-quake/releases/tag/v2.0.0-pre5

This includes initial support for auto-starting apps, which can be done in several ways (see release notes).

**Reply by neondeex** [reply]
Amazing! update the windget version

**Reply by flyingpie** [reply]
@neondeex WinGet release is underway for prerelease 6:
https://github.com/microsoft/winget-pkgs/pull/147158
https://github.com/flyingpie/windows-terminal-quake/releases/tag/v2.0.0-pre6

**Comment by doodead** [comment]
Am I correct that the window size is not configurable yet as of the current state of the reworked v2? It's been the only thing missing for me to consider it working perfectly.

Either way, thank you so much for the great work, man! ❤️

**Reply by flyingpie** [reply]
@mateusmedeiros Thank you for reaching out, and thank you! That is correct, I'll prioritize that as it's small and very useful.

I'll ping you when it's here :)

**Reply by flyingpie** [reply]
@mateusmedeiros Windows resizing has been added, I'd love to hear if this works for you :)
https://github.com/flyingpie/windows-terminal-quake/releases/tag/v2.0.0-pre6

**Reply by doodead** [reply]
Thanks! I've been busy the last few days with IRL stuff, gonna get back to work around friday and will let you know!

**Reply by doodead** [reply]
@flyingpie I've been testing it for the last 2 hours and it's been working perfectly. I tried a bunch of options with the new 3 settings you added and at first glance it worked exactly as expected.

Coupled with my undecorated wezterm I now have a truly fullscreen terminal that I can open and close with a single keypress. ☺️  

Thank you again for all the great work!
❤️

**Comment by PiAlma** [comment]
Great tool!
I have some remarks (for [v2.0.0-prerelease6](https://github.com/flyingpie/windows-terminal-quake/releases/tag/v2.0.0-pre6)):
1. Can I use multiple modifiers on the same shortcut ?
2. I would like to remove the windows from the control of wtq become I temporarily want to keep it on screen.
      - Currently: shortcut display the windows. I resized it. The windows lose focus and disappear.  
      - What I would like: shortcut display the windows. I resized it (hint for wtq to deactivate its control over this window). The window loose focus, but stay visible until I trigger the shortcut again.
3. If wtq  AttachMode is FindOrStart, it will try to start the program, and after a time-out, try again, and again... In my case, I tried the PowerShell from the default wtq.jsonc. But when I launch PowerShell, it is captured by terminal insider. So wtq kept spawning PowerShell. Could you display an error message instead of spawning a new process in case of failure?

Setting to capture PowerShell inside terminal insider:
![image](https://github.com/flyingpie/windows-terminal-quake/assets/152856817/455c586f-ed22-4366-9e65-a6b2685a324c)

**Comment by flyingpie** [comment]
@PiAlma Thank you for your feedback, I really appreciate it! :)

1. You mean like Ctrl+Shift+Q or something?
If that's the case, yes, I've _just_ written some docs about this here: http://localhost:8000/v2/settings/hotkeys/

TL;DR: List the keys with a comma in between, like:

```json
{
    "HotKeys": [
        { "Modifiers": "Control,Shift", "Key": "W" }
    ]
}
```

2. I totally get what you're saying, and I'm also in favor of such a feature. The main thing that I'm not quite sure of, is the hint that wtq needs to receive so it knows that it should stop grabbing the app for a bit. I was considering taking a look at the app location, and see if it differs much from where wtq last left it. Do you have any thoughts on this?

3. Interesting one! I did notice that the constant retrying doesn't work great, so it's definitely an area that needs work. A message box to tell you what's going on is very welcome I feel, cause now you're being Jon Snow'ed. I was also wondering about only having the find-or-start process run when wtq starts (like maybe a couple times in the name of retry), then show a message if it fails. Then maybe try again when you actually hit the shortcut.

**Reply by PiAlma** [reply]
1. It is perfect, that was what I mean. 
2. App location is a good idea. You could also add dimension to detect window resizing.

Thank you again for this great tool.

**Comment by Heratiki** [comment]
I'm not sure what I've done wrong, but I don't believe wtq is working for me at all.  If I'm running the Terminal and try and run wtq it just goes back to an empty prompt, and nothing happens.  I've attempted installing via winget, winget prerelase, and scoop but still it doesn't do anything.  If I run it via the Run Dialog it starts as a background process and just keeps opening new Terminal Windows (none of which are docked).  Hoping someone can help.

**Reply by flyingpie** [reply]
Hey @Heratiki!

I've added some suggestions here: https://github.com/flyingpie/windows-terminal-quake/issues/130

Would you want to take a look and see what happens?

**Comment by neondeex** [comment]
Is there a way to make transparent every window transparent not only quake?
![image](https://github.com/flyingpie/windows-terminal-quake/assets/30392593/23f8c15e-1991-406e-b032-50b4fed64a13)
Also in my portrait screen the quake does not work right

**Reply by flyingpie** [reply]
@neondeex Could you show me your wtq config file?

**Reply by neondeex** [reply]
```{
  // V2 WIP Discussion: https://github.com/flyingpie/windows-terminal-quake/discussions/119
  // Docs: https://wtq.flyingpie.nl

  "Apps": [
    // Example with PowerShell
    {
      "Name": "Terminal",
      "HotKeys": [{ "Modifiers": "Control", "Key": "D1" }],
      "FileName": "WindowsTerminal",
      "AttachMode": "Find",
      "ProcessName": "WindowsTerminal",
      "PreferMonitor": "AtIndex",
      "MonitorIndex": 0,
      "VerticalOffset": 0,
      "Opacity": 95,
      "HorizontalScreenCoverage": 60,
      "VerticalScreenCoverage": 80
    },
    {
      "Name": "Notion",
      "HotKeys": [{ "Modifiers": "Control", "Key": "D3" }],
      "AttachMode": "Find",
      "FileName": "Notion",
      "PreferMonitor": "AtIndex",
      "MonitorIndex": 0,
      "Opacity": 98,
      "VerticalOffset": 0,
      "HorizontalScreenCoverage": 70,
      "VerticalScreenCoverage": 90
    },
    {
      "Name": "Q-Dir",
      "HotKeys": [{ "Modifiers": "Control", "Key": "D2" }],
      "AttachMode": "Find",
      "FileName": "Q-Dir",
      "PreferMonitor": "AtIndex",
      "MonitorIndex": 0,
      "Opacity": 90,
      "VerticalOffset": 0,
      "HorizontalScreenCoverage": 60,
      "VerticalScreenCoverage": 80
    },
    {
      "Name": "Spotify",
      "HotKeys": [{ "Modifiers": "Control", "Key": "D4" }],
      "FileName": "Spotify",
      "PreferMonitor": "AtIndex",
      "MonitorIndex": 0,
      "VerticalOffset": 100,
      "HorizontalScreenCoverage": 60,
      "VerticalScreenCoverage": 80
    },
    {
      "Name": "Figma",
      "HotKeys": [{ "Modifiers": "Control", "Key": "D5" }],
      "FileName": "Figma",
      "PreferMonitor": "AtIndex",
      "MonitorIndex": 1,
      "VerticalOffset": 0,
      "HorizontalScreenCoverage": 100,
      "VerticalScreenCoverage": 100
    }
  ],

  // Hot keys that toggle the most recent app (or the first one if none has been active yet).
  "HotKeys": [{ "Modifiers": "Control", "Key": "Q" }]

  // "WithCursor" (default), "Primary" or "AtIndex".
}
```

**Reply by flyingpie** [reply]
@neondeex Thank you for the config!

So if I understand you correctly, you want this app to be transparent, correct?

![image](https://github.com/flyingpie/windows-terminal-quake/assets/1295673/a8d1c80c-ad89-4bdd-926b-d357b64c6ca9)

If so, could you tell me what app it is?

**Reply by flyingpie** [reply]
@neondeex Do you mean you want to set transparency for apps that are not managed by wtq, and not toggled at all?

**Reply by flyingpie** [reply]
@neondeex Regarding this part:
> Also in my portrait screen the quake does not work right

Could you elaborate on that? I am aware of an issue where apps are not going off the screen entirely, if you have multiple screens where the primary is lower than the secondary, like so:

![image](https://github.com/flyingpie/windows-terminal-quake/assets/1295673/6709543a-4731-4006-ac7b-ac1703444584)

Is it the same one, or something else?

**Reply by neondeex** [reply]
> gled

Yes thats what i mean
yeah thats my issue

**Reply by flyingpie** [reply]
@neondeex Thanks!

Regarding the issue around multi-monitor, I've actually starting using a vertical setup myself, and expect to have a fix for this within a couple of days.

Regarding the transparency, if the app is not intended to be toggle on/off screen (eg. the whole point of WTQ), I'm not sure it _should_ be a feature per se.

Though, I think that with the next release, it _should_ be possible to set the parameters of a WTQ app such, that it won't move the app, but it will apply transparency. I'll get back to you when it's ready.

**Reply by flyingpie** [reply]
@neondeex The patch for the vertical monitor setup is ready, can you take a look here?

https://github.com/flyingpie/windows-terminal-quake/releases/tag/v2.0.10

**Reply by neondeex** [reply]
Please update the scoop version
![image](https://github.com/user-attachments/assets/a1ac9580-884f-4a62-b5ca-26665fa37164)

**Reply by flyingpie** [reply]
@neondeex Updated!

**Reply by neondeex** [reply]
Amazing! All the new updates improved my workflow a loot and now its an amazing app

**Reply by flyingpie** [reply]
@neondeex Thank you for your feedback! :)

**Comment by timorbit** [comment]
Hi, I'm loving wtq! I've been trying to get it to work properly with Whatsapp Desktop, but when Whatsapp is in it's 'passive' mode, there's always a white canvas covering the screen. Is this an issue other apps have too? And if so, is there a workaround?

Thanks for your work!

**Reply by flyingpie** [reply]
@Bobjeslol Thank you for your kind words, and for reaching out! :heart: 

Could you post a (crossed out) screenshot of what you're seeing? I'll fire up a Windows machine to test it on my end meanwhile.

**Reply by timorbit** [reply]
A restart completely changed its behavior. It still isn't doing what I want it to do, though.

Currently the assigned hotkey just opens whatsapp in fullscreen as if i launched the program normally. The swap button doesn't work for whatsapp, it only works for the other apps i have assigned.

This is my whatsapp entry in the settings file:
`
		{
			"Name": "Whatsapp",
    			"Hotkeys": [{ "Modifiers": "Control", "Key": "D3" }],
    			"FileName": "C:\\Program Files\\WindowsApps\\5319275A.WhatsAppDesktop_2.2445.7.0_x64__cv1g1gvanyjgm\\WhatsApp.exe",
    			"ProcessName": "WhatsApp"
		}
`

I'll try reinstalling whatsapp via the actual executable instead of via the microsoft store, that might make things easier!

**Reply by flyingpie** [reply]
@Bobjeslol So this seems to have something to do with the sandboxing that's going on with MS Store apps.

I've noticed that they're making it hella hard to do it the "old" way, but let me know if that works!

I'll make an issue for MS Store apps, no promises though, given that the whole point is to keep apps isolated from one another.

**Reply by timorbit** [reply]
Yeah, that seems to be the issue. Can't get it to work, so I just decided to leave whatsapp out of it. It's time to move to Signal anyways. Thanks for your help!
